keps/sig-architecture/4330-compatibility-versions/README.md

```diff
@@ -0,0 +1,1106 @@
+
+```go
+map[Feature]VersionedSpecs{
+		featureA: VersionedSpecs{
+			{Version: mustParseVersion("1.26"), Default: false, PreRelease: Alpha},
+			{Version: mustParseVersion("1.27"), Default: true, PreRelease: Beta},
+			{Version: mustParseVersion("1.28"), Default: true, PreRelease: GA},
+		},
+}
+```
+
+The lifecycle of the feature would be:
+
+| Release | Stage | Feature tracking information                      |
+| ------- | ----- | ------------------------------------------------- |
+| 1.26    | alpha | Alpha: 1.26                                       |
+| 1.27    | beta  | Alpha: 1.26, Beta: 1.27 (on-by-default)           |
+| 1.28    | GA    | Alpha: 1.26, Beta: 1.27 (on-by-default), GA: 1.28 |
+| 1.29    | GA    | Alpha: 1.26, Beta: 1.27 (on-by-default), GA: 1.28 |
+| 1.30    | GA    | Alpha: 1.26, Beta: 1.27 (on-by-default), GA: 1.28 |
+| 1.31    | GA    | **Feature implementation becomes part of normal code, Feature gate and feature tracking information may be removed from code** |
+
+All feature gating and tracking must remain in code through 1.30 for N-3
+compatibility version support.
+
+For a feature that is removed, e.g.:
+
+```go
+map[Feature]VersionedSpecs{
+		featureA: VersionedSpecs{
+			{Version: mustParseVersion("1.26"), Default: false, PreRelease: Alpha},
+			{Version: mustParseVersion("1.27"), Default: false, PreRelease: Deprecated},
+			{Version: mustParseVersion("1.31"), Default: false, PreRelease: Removed},
+		},
+}
+```
+
+The lifecycle of the feature implementation would be:
+
+| Release | Stage | Feature tracking information                      |
+| ------- | ----- | ------------------------------------------------- |
+| 1.26    | alpha | Alpha: 1.26                                       |
+| 1.27    | alpha | Alpha: 1.26, Deprecated: 1.27                     |
+| 1.28    | alpha | Alpha: 1.26, Deprecated: 1.27                     |
+| 1.29    | alpha | Alpha: 1.26, Deprecated: 1.27                     |
+| 1.30    | -     | Alpha: 1.26, Deprecated: 1.27, Removed: 1.31      |
+| 1.31    | -     | Alpha: 1.26, Deprecated: 1.27, Removed: 1.31      |
+| 1.32    | -     | Alpha: 1.26, Deprecated: 1.27, Removed: 1.31      |
+| 1.33    | -     | **Feature implementation and feature tracking information may be removed from code** |
+
+Note that this respects a 1yr deprecation policy.
+
+All feature gating and tracking must remain in code through 1.32 for N-3
+compatibility version support.
+
+#### Feature gating changes
+
+In order to preserve the behavior of in-development features across multiple releases,
+feature implementation history should also be preserved in the code base.
+
+Only sigificant and observable changes in feature capabilities should be across
+releases. We do not want to impose a unreasonable burdon on feature authors.
+Bugs, performance optimizations should not be gated by version. 
+
+Naively, the feature implementations can be gated by version number. 
```

Comment:
This feels like quite a lot to expect developers to get right.  It's already tricky to review logic that spans alpha/beta (`if gateEnabled(foo) || fooInUse(obj)`).  I think we would a) need to establish very clear criteria for when this is and is not necessary (and bias towards "less work"); and b) document or even write helper libs to make this as low-touch as possible.

Reply:
I think one way to reduce developer burden is to make it clear compatibility version does not provide any promises for alpha features, since alpha features can still be under active development. 
For beta+ features, they should be more or less stable. And only user-facing api/function/breaking changes should be gated by version.

Reply:
I really like the no comparability version support for Alpha idea. That simplifies things dramatically.

Reply:
I was doing a KEP update so I added the "compatibility version may not be used in conjunction with alpha features/APIs" idea. PTAL

Reply:
I don't think the feature gate API with compat version would explicitly not work with alpha features/APIs. My understanding is that there is no guarantee that alpha features would work consistently for the same compat version across different binary versions.

Reply:
I still think we need REALLY clear criteria and examples of when and how this is to be done.

Reply:
Agreed. Let's make this crisp.

My concern with allowing compat version in conjunction with alpha features is that it requires extra work.  Take the case of a feature introduced as alpha and then later deleted.   If we want compat version to work at all, we need to keep the alpha code around for the n-3 version range before actually deleting it.

Reply:
@siyuanfoundation added some examples: https://github.com/kubernetes/enhancements/pull/4395/commits/637db38f1b10319a12cd9e4eb237a5da8bfe1314

Feedback welcome.